### DISCUSSION TITLE
Hitsamples in keysounded maps seen to be wrongly assigned

### DISCUSSION THREAD

**Discussion by MaxKruse** [ppy/osu#30437] [Q&A] [answered]
I saw that https://github.com/ppy/osu/issues/24997 got closed and for good reason, that case was closed, so i decided to double-check some maps i liked and see if i can find anything else. Turns out, i can:

For comparison sake, im using the 2013 skin in lazer, as well as the default in "normal" osu! (the same skin) because as far as im concerned, every single map that was hitsounded to this day and doesnt fully replace all hitsamples was built on the sounds of the 2013 skin's included samples.

On the map https://osu.ppy.sh/beatmapsets/1633250#osu/3333745 the following section 02:38:227 (1) - has a very distinctly wrong sample assignment, a video to illustrate this: (apologies for the recording + reencoding framerate+quality, the audio however is unaffected)

Lazer: 

https://github.com/user-attachments/assets/f75a579d-3310-4365-b920-2a964219cd6d

osu!: 

https://github.com/user-attachments/assets/6358140c-ee5c-46dc-901b-d1325ba062e4

**Comment by peppy** [comment]
For clarity's sake, the beatmap you linked is not "keysounded", ie. not using per-object assignments. Hit sounds are being assigned using control points with custom banks.

Based on the videos, a point of concern is hitcircle:

`237,158,158227,5,2,3:0:0:0:`

affected by control point:

`157937,-62.5,4,2,7,65,0,0`

The hitobject does seem to be getting the correct definition:

![JetBrains Rider-EAP 2024-11-13 at 07 08 16](https://github.com/user-attachments/assets/71e5df94-b4a6-4df9-9ca6-719f959e77cc)

which leads to lookup names for the sample:

![JetBrains Rider-EAP 2024-11-13 at 07 11 30](https://github.com/user-attachments/assets/dc647f01-491a-4515-b6dd-aa72366099a3)

succeeds on `drum-hitwhistle7`, which does exist in the beatmap:

![JetBrains Rider-EAP 2024-11-13 at 07 12 12](https://github.com/user-attachments/assets/ca273b8e-aade-471f-8bf5-772c6fb643e0)

So far everything looks fine. In fact, if you listen closely the sample is actually playing. It's just drowned up by `drum-hitnormal` also playing. Can be confirmed with this patch:

```diff
diff --git a/osu.Game/Skinning/PoolableSkinnableSample.cs b/osu.Game/Skinning/PoolableSkinnableSample.cs
index 76c2c4d7ec..9ee4b7c371 100644
--- a/osu.Game/Skinning/PoolableSkinnableSample.cs
+++ b/osu.Game/Skinning/PoolableSkinnableSample.cs
@@ -104,7 +104,8 @@ private void updateSample()
             if (sample == null)
                 return;
 
-            sampleContainer.Add(Sample = new DrawableSample(sample));
+            if (!sample.Name.Contains("hitnormal"))
+                sampleContainer.Add(Sample = new DrawableSample(sample));
 
             // Start playback internally for the new sample if the previous one was playing beforehand.
             if (wasPlaying && Looping)

```

The `drum-hitnormal` which is playing is included in the beatmap, but without a custom prefix:

![Finder 2024-11-13 at 07 34 42](https://github.com/user-attachments/assets/5e567a3b-d73d-4b47-9b6f-16cae4902b2c)

In conclusion, I think the fallback logic we have doesn't match stable. If a custom suffix/sampleset number is specified and doesn't exist, the expectation seems to be "play nothing".

Checking stable's code:

[Playback always prefers the cache, even when the lookup is cached as a failure `-1`](https://github.com/peppy/osu-stable-reference/blob/7519cafd1823f1879c0d9c991ba0e5c7fd3bfa02/osu!/Audio/AudioEngine.cs#L1208-L1213)

[Cache / load code starts here; follow the chain](https://github.com/peppy/osu-stable-reference/blob/bb57924c1552adbed11ee3d96cdcde47cf96f2b6/osu!/GameplayElements/HitObjectManager_LoadSave.cs#L127)

**Reply by peppy** [reply, marked as answer]
Should be resolved via https://github.com/ppy/osu/pull/30599.
